### PR TITLE
Feature/carma wm ctrl

### PR DESCRIPTION
**Pull request by msmcconnell**
# PR Details
## Description

This pr adds the carma_wm_ctrl package which contains a world model broadcaster node. This node is meant to subscribe to the loaded lanelet2 map and to incoming geofence data. When a new geofence is added. It will update the lanelet2 map and broadcast a new map. Since the geofence spec is not finalized and the best methodology for updating lanelet maps has not yet been determined the focus of this version was on geofence scheduling. A geofence schedule is currently planned to be defined as follows 

	<schedule start="1970-01-01T00:00:00Z" end="9999-12-31T23:59:59Z">
		<dow su="" mo="true" tu="" we="" th="" fr="" sa=""/>
		<between start="15:00:00-5:00" end="17:30:00-5:00"/>
		<repeat interval="60" duration="30" uom="min"/>
	</schedule>
There is a overall schedule start and end time. During this time the schedule can repeat during different days of the week. Each day it will start at start time then run for one "duration" then wait till one "interval" since the last start time has passed. Then it continue for another duration until the daily end time is reached. There are some edge cases like local time (instead of UTC) and crossing midnight, that are not handled in this PR. These are left for future work. 

The class structure is as follows 
Geofence.h : A basic geofence object. Contains only a schedule and id at this time
GeofenceSchedule.h : A schedule which a geofence adheres too
GeofenceScheduler.h : A multi-threaded scheduler which uses timers to trigger callbacks when a geofence goes active or inactive. 
Timer.h : A timer interface used to do dependency injection for unit tests
TimerFactory.h : A timer factory interface used for dependency injection in unit tests of the scheduler
ROSTimer.h An implementation of Timer.h using ros node handle and ros::Timer
ROSTimerFactory.h An implementation of TimerFactory for creating ROSTimers
WMBroadcaster.h The worker class for implementing map updating and reacting to geofences
WMBroadcasterNode.h The node class which connected to ros
MapConformer.h Functions which can be used to add carma regulatory elements to maps which match the lanelet2 tagging standard. 

This PR also adds a notice on carma_wm about the added regulatory elements. 

The unit tests also contain a TestTimer.h and TestTimerFactory.h which are custom timers built with std::thread and the ros::Time/Duration classes. 
## Related Issue

#441
## Motivation and Context

Implement initial logic for world model updates
## How Has This Been Tested?

Integration tested on the blue lexus
## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: carma_wm_ctrl/src/WMBroadcaster.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  scheduler_.addGeofence(gf);  // Add the geofence to the schedule
+  ROS_INFO_STREAM("New geofence message received by WMBroadcaster with id" << gf.id_);
+};
+
+void WMBroadcaster::addGeofence(const Geofence& gf)
+{
+  std::lock_guard<std::mutex> guard(map_mutex_);
+  ROS_INFO_STREAM("Adding active geofence to the map with geofence id: " << gf.id_);
+  // TODO Add implementation for adding a geofence
```
Comment (qswawrq) [inline carma_wm_ctrl/src/WMBroadcaster.cpp]: Do we need to complete this function and removeGeofence in this PR for it is intentional left as blank?

**Comment by qswawrq** [comment]
Looks like this branch cannot be built without a new Autoware image?

File: carma_wm_ctrl/include/carma_wm_ctrl/Timer.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+#include <ros/time.h>
+#include <ros/timer.h>
+
+namespace carma_wm_ctrl
+{
+/**
+ * @brief A timer class interface which will trigger the provided callback after the requested duration
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply carma_wm_ctrl/include/carma_wm_ctrl/Timer.h]: No unfortunately creating a ros::Timer requires a node handle. So to use it in a unit test I need this interface.

File: carma_wm_ctrl/include/carma_wm_ctrl/GeofenceScheduler.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  uint32_t next_id_ = 0;  // Timer id counter
+
+public:
+  /**
+   * @brief Constructor which takes in a TimerFactory. Timers from this factory will be used to generate the triggers
+   * for goefence activity.
+   *
+   * @param timerFactory A pointer to a TimerFactory which can be used to generate timers for geofence triggers.
+   */
+  GeofenceScheduler(std::unique_ptr<TimerFactory> timerFactory);
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply carma_wm_ctrl/include/carma_wm_ctrl/GeofenceScheduler.h]: You have to use std::move
GeofenceScheduler(std::move(my_unique))

File: carma_wm_ctrl/src/WMBroadcaster.cpp (hunk 1 above)
Comment (msmcconnell) [reply carma_wm_ctrl/src/WMBroadcaster.cpp]: Intentionally left blank. Actually adding the geofences to the map will take quite a bit of work and is not in the scope of this task.

**Comment by msmcconnell** [comment]
@qswawrq Correct we need a component release of autoware.ai before this can be built

**Comment by msmcconnell** [comment]
I have to update the rosjava mock drivers message version because astuff updated their repo
https://github.com/astuff/astuff_sensor_msgs

**Comment by msmcconnell** [comment]
While our CI is broken, the build passed as did the unit tests for this package.

**Pull request merged**